source/material_model/viscoelastic_plastic.cc

```diff
@@ -0,0 +1,796 @@
+          fe_values[this->introspection().extractors.velocities].get_function_gradients (this->get_old_solution(),
+                                                                                         old_velocity_gradients);
+
+          MaterialModel::ElasticOutputs<dim>
+          *force_out = out.template get_additional_output<MaterialModel::ElasticOutputs<dim> >();
+
+          for (unsigned int i=0; i < in.position.size(); ++i)
+            {
+
+              const std::vector<double> composition = in.composition[i];
+              const std::vector<double> volume_fractions = compute_volume_fractions(composition, composition_mask);
+
+              // Get old stresses from compositional fields
+              SymmetricTensor<2,dim> stress_old;
+              for (unsigned int j=0; j < SymmetricTensor<2,dim>::n_independent_components; ++j)
+                stress_old[SymmetricTensor<2,dim>::unrolled_to_component_indices(j)] = in.composition[i][j];
+
+              // Calculate the rotated stresses
+              // Rotation (vorticity) tensor (equation 25 in Moresi et al., 2003, J. Comp. Phys.)
+              const Tensor<2,dim> rotation = 0.5 * ( old_velocity_gradients[i] - transpose(old_velocity_gradients[i]) );
+
+              // Recalculate average elastic shear modulus
+              const double elastic_shear_modulus = average_value(volume_fractions,
+                                                                 elastic_shear_moduli,
+                                                                 viscosity_averaging);
+
+              // Average viscoelastic viscosity
+              const double viscoelastic_viscosity = out.viscosities[i];
+
+              // Calculate the current (new) viscoelastic stress, which is a function of the material
+              // properties (viscoelastic viscosity, shear modulus), elastic time step size, strain rate,
+              // vorticity and prior (inherited) viscoelastic stresses (see equation 29 in Moresi et al.,
+              // 2003, J. Comp. Phys.)
+              SymmetricTensor<2,dim> stress_new = ( 2. * viscoelastic_viscosity * deviator(in.strain_rate[i]) ) +
+                                                  ( ( viscoelastic_viscosity / ( elastic_shear_modulus * dte ) ) * stress_old ) +
+                                                  ( ( viscoelastic_viscosity / elastic_shear_modulus ) *
+                                                    ( symmetrize(rotation * Tensor<2,dim>(stress_old) ) - symmetrize(Tensor<2,dim>(stress_old) * rotation) ) );
+
+              // Stress averaging scheme to account for difference betweed fixed elastic time step
+              // and numerical time step (see equation 32 in Moresi et al., 2003, J. Comp. Phys.)
+              const double dt = this->get_timestep();
+              if (use_fixed_elastic_time_step == true && use_stress_averaging == true)
```

Comment:
Do you need to check that dte < dt?

Reply:
Good catch! However, I think the check should be that `dt < dte`. If ` dt >>>> dte` it could lead to negative stresses.